### PR TITLE
[Example] Add TGCN Model for Traffic Forecasting

### PR DESCRIPTION
**Pull request by casia-rxwang**
### PR types

Others 

### PR changes

Others  

### Describe

* add TGCN model for traffic forecasting
* add docs for TGCN
* add example for TGCN
* add interface for PEMSD4 & PEMSD8 dataset

### CODE REVIEW

File: docs/zh/examples/tgcn.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
+# TGCN
+
+开始训练、评估前，请下载数据集：[PEMSD4 & PEMSD8](https://paddle-org.bj.bcebos.com/paddlescience/datasets/tgcn/tgcn_data.zip)。将解压后的数据集文件夹与 `PaddleScience` 文件夹放置于同一目录下。
+
+开始评估前，请下载或训练生成预训练模型：[PEMSD4](https://paddle-org.bj.bcebos.com/paddlescience/models/tgcn/PEMSD4_pretrained_model.pdparams) & [PEMSD8](https://paddle-org.bj.bcebos.com/paddlescience/models/tgcn/PEMSD8_pretrained_model.pdparams)。将预训练模型文件与 `PaddleScience` 文件夹放置于同一目录下。
+
```
Comment (HydrogenSulfate) [inline docs/zh/examples/tgcn.md:6]: 3-5行的说明可以删除，训练和评估开始前的数据集下载命令请直接放到文档的命令中，参考如下
image
image

File: docs/zh/examples/tgcn.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+## 5. 结果展示
+
+下表展示了 TGCN 在 PEMSD4 和 PEMSD8 两个数据集上的评估结果。
+
+| 数据集 | MAE   | RMSE  |
+| :----- | :---- | :---- |
+| PEMSD4 | 21.48 | 34.06 |
+| PEMSD8 | 15.57 | 24.52 |
+
```
Comment (HydrogenSulfate) [inline docs/zh/examples/tgcn.md:295]: 文档开头位置加上数据集和对应的预训练模型，如下所示

image

File: examples/tgcn/conf/run.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    config:
+      override_dirname:
+        exclude_keys:
+          - mode
+          - output_dir
+          - log_freq
+  sweep:
+    # output directory for multirun
+    dir: ${hydra.run.dir}
+    subdir: ./
```
Comment (HydrogenSulfate) [inline examples/tgcn/conf/run.yaml:17]:
```suggestion
defaults:
  - ppsci_default
  - TRAIN: train_default
  - TRAIN/ema: ema_default
  - TRAIN/swa: swa_default
  - EVAL: eval_default
  - INFER: infer_default
  - _self_

hydra:
  run:
    # dynamic output directory according to running time and override name
    dir: outputs_tgcn/${now:%Y-%m-%d}/${now:%H-%M-%S}
  job:
    name: ${mode} # name of logfile
    chdir: false # keep current working directory unchanged
  callbacks:
    init_callback:
      _target_: ppsci.utils.callbacks.InitCallback
  sweep:
    # output directory for multirun
    dir: ${hydra.run.dir}
    subdir: ./
```

File: examples/tgcn/conf/run.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+norm_input: True
+norm_label: False
+reduce: mean
+
+# model settings
+MODEL:
+  afno:
+    input_keys: ['input']
+    label_keys: ['label']
+
```
Comment (HydrogenSulfate) [inline examples/tgcn/conf/run.yaml:40]: 单个模型可以删掉`afno:`

```suggestion
MODEL:
  input_keys: ['input']
  label_keys: ['label']

```

File: examples/tgcn/conf/run.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+# training settings
+TRAIN:
+  epochs: 200
+  learning_rate: 0.01
+  pretrained_model_path: null
+  batch_size: ${batch_size}
+
+# evaluation settings
+EVAL:
+  pretrained_model_path: ${data_name}_best.pdparams
```
Comment (HydrogenSulfate) [inline examples/tgcn/conf/run.yaml:63]: 配置中请设置为null，在评估命令中再指定路径

File: examples/tgcn/run.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    logger.message(cfg)
+
+    # set train dataloader config
+    train_dataloader_cfg = {
+        'dataset': {
+            'name': 'PEMSDataset',
+            'file_path': cfg.data_path,
+            'split': 'train',
+            'input_keys': cfg.MODEL.afno.input_keys,
+            'label_keys': cfg.MODEL.afno.label_keys,
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:26]: 全局搜索修改一下：`cfg.MODEL.afno.` --> `cfg.MODEL.`

File: examples/tgcn/run.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+def eval(cfg: DictConfig):
+    # set random seed for reproducibility
+    ppsci.utils.misc.set_random_seed(cfg.seed)
+
+    # initialize logger
+    logger.init_logger('ppsci', os.path.join(cfg.output_dir, 'test.log'), 'info')
+    logger.message(cfg)
+
+    # set eval dataloader config
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:122]:
```suggestion
    # set eval dataloader config
```

File: examples/tgcn/run.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+def train(cfg: DictConfig):
+    # set random seed for reproducibility
+    ppsci.utils.misc.set_random_seed(cfg.seed)
+
+    # initialize logger
+    logger.init_logger('ppsci', os.path.join(cfg.output_dir, 'train.log'), 'info')
+    logger.message(cfg)
+
+    # set train dataloader config
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:19]:
```suggestion
    # set train dataloader config
```

File: ppsci/arch/tgcn.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        tc_layer = cfg.tc_layer
+        k_s = cfg.tc_kernel_size
+        dropout = cfg.dropout
+        alpha = cfg.leakyrelu_alpha
+
+        self.input_keys = cfg.MODEL.afno.input_keys
+        self.output_keys = cfg.MODEL.afno.label_keys
+
+        self.edge_index = pp.to_tensor(data=edge_index, place=cfg.device)
+        self.edge_attr = pp.to_tensor(data=edge_attr, place=cfg.device)
```
Comment (HydrogenSulfate) [inline ppsci/arch/tgcn.py:71]: 这两个属性好像并没有被用到？

File: ppsci/arch/tgcn.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        self.adj = pp.to_tensor(data=adj, place=cfg.device)
```
Comment (HydrogenSulfate) [inline ppsci/arch/tgcn.py:72]: 1. 跟pytorch不同，to_tensor的place不需要手动指定，会默认使用当前全局设备设置
2. 在__init__中生成的张量，如果不需要进行梯度更新，则建议使用如下方式注册为 buffer，

image

如需要梯度更新，则注册为parameter，
image

否则模型保存的时候，不会将这些变量持久化保存到参数文件中

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

**Comment by HydrogenSulfate** [comment]
@casia-rxwang  签署一下CLA协议：
image

**Comment by HydrogenSulfate** [comment]
@casia-rxwang 修改代码的时候顺便合并一下最新develop分支的代码，并解决一下冲突

File: examples/tgcn/run.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        device=cfg.device,
+        validator=validator,
+        pretrained_model_path=cfg.TRAIN.pretrained_model_path,
+        # compute_metric_by_batch=False,
+        eval_with_no_grad=True,
+    )
+    # train model
+    solver.train()
+    # evaluate after training
+    eval(cfg)
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:103]: 这里是不是调solver.eval()就行了呀

File: ppsci/data/dataset/pems_dataset.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    elif reduce == "mean":
+        adj[edge_index[0], edge_index[1]] = 1.0
+        adj = adj / adj.sum(axis=-1)
+    else:
+        raise ValueError
+
+    return edge_index, edge_attr, adj
+
+
+class PEMSDataset(Dataset):
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/pems_dataset.py:67]: 这个公开的API请补充一下docstring，并添加到docs/zh/api/data/dataset.md中

File: ppsci/arch/tgcn.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        # B C N T
+        x_cat = [x]
+        for i in range(self.num_layer):
+            x = self.leakyrelu(self.tc_convs[i](x))
+            x_cat.append(x)
+        tc_out = self.mlp(pp.concat(x_cat, axis=1))
+        return tc_out
+
+
+class TGCN(Arch):
```
Comment (HydrogenSulfate) [inline ppsci/arch/tgcn.py:71]: 这个公开的API请补充一下docstring，并添加到docs/zh/api/arch.md中

File: docs/zh/examples/tgcn.md [hunk 14]
```diff
@@ -0,0 +1,284 @@
+# TGCN
+
+=== "模型训练命令"
+
+    ``` sh
+    # Train
```
Comment (HydrogenSulfate) [inline docs/zh/examples/tgcn.md:6]:
```suggestion
```

File: docs/zh/examples/tgcn.md [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+    wget https://paddle-org.bj.bcebos.com/paddlescience/datasets/tgcn/tgcn_data.zip
+    unzip tgcn_data.zip
+    python run.py data_name=PEMSD8
+    # python run.py data_name=PEMSD4
+    ```
+
+=== "模型评估命令"
+
+    ``` sh
+    # Eval
```
Comment (HydrogenSulfate) [inline docs/zh/examples/tgcn.md:16]:
```suggestion
```

File: examples/tgcn/run.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+            "norm_input": cfg.norm_input,
+            "norm_label": cfg.norm_label,
+            "input_len": cfg.input_len,
+            "label_len": cfg.label_len,
+        },
+        "sampler": {
+            "name": "BatchSampler",
+            "drop_last": False,
+            "shuffle": False,
+        },
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:54]: 如果`drop_last`、`shuffle`都是False，可以直接删除
```suggestion
```

File: examples/tgcn/run.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:128]: 同，可以直接删除
```suggestion
```

File: examples/tgcn/run.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    solver = ppsci.solver.Solver(
+        model=model,
+        output_dir=cfg.output_dir,
+        log_freq=cfg.log_freq,
+        seed=cfg.seed,
+        device=cfg.device,
+        validator=validator,
+        pretrained_model_path=cfg.EVAL.pretrained_model_path,
+        eval_with_no_grad=True,
+    )
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:170]:
```suggestion
    solver = ppsci.solver.Solver(
        model=model,
        output_dir=cfg.output_dir,
        log_freq=cfg.log_freq,
        device=cfg.device,
        validator=validator,
        pretrained_model_path=cfg.EVAL.pretrained_model_path,
        eval_with_no_grad=True,
    )
```

File: examples/tgcn/run.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+    solver = ppsci.solver.Solver(
+        model=model,
+        constraint=constraint,
+        output_dir=cfg.output_dir,
+        optimizer=optimizer,
+        epochs=cfg.TRAIN.epochs,
+        iters_per_epoch=iters_per_epoch,
+        log_freq=cfg.log_freq,
+        eval_during_train=True,
+        seed=cfg.seed,
```
Comment (HydrogenSulfate) [inline examples/tgcn/run.py:100]:
```suggestion
```

File: ppsci/arch/tgcn.py [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+        ...    k_s=3,
+        ...    dropout=0.25,
+        ...    alpha=0.1,
+        ...    input_len=12,
+        ...    label_len=12,
+        ... )
+        >>> input_dict = {"input": paddle.rand([64, 12, 307, 1]),}
+        >>> label_dict = model(input_dict)
+        >>> print(label_dict["label"].shape)
+        [64, 12, 307, 1]
```
Comment (HydrogenSulfate) [inline ppsci/arch/tgcn.py:114]: 缩进少了个空格
```suggestion
        >>> import paddle
        >>> import ppsci
        >>> model = ppsci.arch.TGCN(
        ...     input_keys=("input",),
        ...     output_keys=("label",),
        ...     adj=numpy.ones((307, 307), dtype=numpy.float32),
        ...     in_dim=1,
        ...     emb_dim=32
        ...     hidden=64,
        ...     gc_layer=2,
        ...     tc_layer=2
        ...     k_s=3,
        ...     dropout=0.25,
        ...     alpha=0.1,
        ...     input_len=12,
        ...     label_len=12,
        ... )
        >>> input_dict = {"input": paddle.rand([64, 12, 307, 1]),}
        >>> label_dict = model(input_dict)
        >>> print(label_dict["label"].shape)
        [64, 12, 307, 1]
```

File: ppsci/data/dataset/pems_dataset.py [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+        self.label_keys = label_keys
+        self.weight_dict = weight_dict
+
+        self.transforms = transforms
+        self.norm_input = norm_input
+        self.norm_label = norm_label
+
+        data = np.load(os.path.join(file_path, "{}.npy".format(split))).astype(
+            np.float32
+        )
```
Comment (HydrogenSulfate) [inline ppsci/data/dataset/pems_dataset.py:117]:
```suggestion
        data = np.load(os.path.join(file_path, f"{split}.npy")).astype(
            np.float32
        )
```

**Review by HydrogenSulfate — COMMENTED** [review]
辛苦提交代码，基本上没什么问题了，有几点小的修改建议

**Review by HydrogenSulfate — APPROVED** [review]
辛苦提交PR，看起来没有什么问题了，我们内部测试之后会尽快合入~
